File: GGAndroid_1.0/GGLibrary/src/main/java/io/ganguo/library/ui/extend/BaseDialog.java
package io.ganguo.library.ui.extend;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.View;

import io.ganguo.library.BaseContext;
import io.ganguo.library.R;
import io.ganguo.library.core.event.BusProvider;

/**
 * 对话框 - 基类
 * <p/>
 * Created by zhihui_chen on 14-9-9.
 */
public abstract class BaseDialog extends Dialog implements IContext {

    public BaseDialog(Context context) {
        this(context, R.style.customDialog);
    }

    public BaseDialog(Context context, int theme) {
        super(context, theme);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        beforeInitView();
        initView();
        initListener();
        initData();
    }

    @Override
    protected void onStart() {
        super.onStart();

        // 返回通过Action
        View actionBack = findViewById(R.id.action_back);
        if (actionBack != null) {
            actionBack.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    dismiss();
                }
            });
        }
    }

    @Override
    protected void onStop() {
        super.onStop();

        BusProvider.getInstance().unregister(this);
    }

    @Override
    public void onAttachedToWindow() {
        super.onAttachedToWindow();

        BusProvider.getInstance().register(this);
    }

    @Override
    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        BusProvider.getInstance().unregister(this);
    }

    /**
     * 对话框大小
     *
     * @param width
     * @param height
     */
    public void setSize(int width, int height) {
        getWindow().setLayout(width, height);
    }

    /**
     * 获取上下文环境
     *
     * @return
     */
    @Override
    public <T extends BaseContext> T getAppContext() {
        return BaseContext.getInstance();
    }

    /**
     * 加载layout xml
     */
    protected abstract void beforeInitView();

    /**
     * 加载UI
     */
    protected abstract void initView();

    /**
     * 监听控件
     */
    protected abstract void initListener();

    /**
     * 加载网络数据
     */
    protected abstract void initData();
}

File: GGAndroid_3.0/core/support/core/src/main/java/io/ganguo/core/ui/dialog/BaseDialog.kt
package io.ganguo.core.ui.dialog

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.view.Gravity
import android.view.WindowManager
import io.ganguo.core.R

/**
 * <pre>
 *     author : leo
 *     time   : 2019/10/11
 *     desc   : Dialog - 基类
 * </pre>
 */
abstract class BaseDialog @JvmOverloads constructor(context: Context, theme: Int = R.style.Dialog_Immersion) : Dialog(context, theme), InitResources {
    override fun onCreate(savedInstanceState: Bundle) {
        super.onCreate(savedInstanceState)
        // init resources
        beforeInitView()
        initView()
        initListener()
        initData()
        initAnimation()
        initDialog()
    }


    /**
     * Dialog init
     *
     * @return
     */
    protected open fun initDialog() {
        window!!.setLayout(dialogWidth, dialogHeight)
    }

    /**
     * 弹窗宽度
     *
     * @return
     */
    protected abstract val dialogWidth: Int

    /**
     * 弹窗高度
     *
     * @return
     */
    protected abstract val dialogHeight: Int

    /**
     * 设置弹窗动画
     */
    protected open fun initAnimation() {
        val params = window!!.attributes
        params.width = WindowManager.LayoutParams.MATCH_PARENT
        params.height = WindowManager.LayoutParams.WRAP_CONTENT
        params.windowAnimations = R.style.Dialog_Animation
        params.gravity = Gravity.BOTTOM
        window!!.attributes = params
    }
}